spinalcordtoolbox/deepseg/models.py

```diff
@@ -0,0 +1,57 @@
+# coding: utf-8
+"""
+List of available models.
+"""
+
+
+import os
+import logging
+
+from spinalcordtoolbox import __sct_dir__
+
+logger = logging.getLogger(__name__)
+
+
+MODELS = {
+    'cord-t2star':
+        {'url': 'https://osf.io/v9hs8/download?version=1',
+         'description': 'Cord segmentation on T2*-weighted contrast.'},
+    'uqueensland-mice-sc':
+        {'url': 'https://osf.io/nu3ma/download?version=1',
+         'description': 'Cord segmentation on mouse MRI. Data from University of Queensland.'},
+    'uqueensland-mice-gm':
+        {'url': 'https://osf.io/mfxwg/download?version=1',
+         'description': 'Gray matter segmentation on mouse MRI. Data from University of Queensland.'},
+    }
+
+
+class DeepsegModel:
+    """
+    Main class that deals with deep learning models. When initialized, the class will inherit the subfields from
+    MODELS[modelx] dict.
+    """
+    def __init__(self, name_model):
+        self.name = name_model
+        self.folder = os.path.join(__sct_dir__, 'models', name_model)
+        for key in MODELS[name_model]:
+            setattr(self, key, MODELS[name_model][key])
+
+    def is_installed(self):
+        """
+        Check if model (.pt file) is installed under SCT directory
+        """
+        if os.path.exists(os.path.join(self.folder, self.name + '.pt')) and \
+                os.path.exists(os.path.join(self.folder, self.name + '.json')):
+            return True
+        else:
+            raise FileNotFoundError("The model is not properly installed. Both the .pt and .json files should be "
+                                    "present, and the basename should be the same as the folder name. Example: "
+                                    "my_model/my_model.pt, my_model/my_model.json")
+
+    def install(self):
+        """
+        Download and install model under SCT directory
+        :param name_model:
+        :return:
+        """
+        raise NotImplementedError
```

Comment:
I think it's simpler to keep the data all together in one place:

```suggestion
class DeepsegModel:
    """
    Metadata of packaged deep learning models.
    """
    def __init__(self, name, url, description):
        self.name = name    # this boilerplate could be removed with py3.7's @dataclass
        self.url = url
        self.description = description

    @property
    def folder(self):
        return os.path.join(__sct_dir__, 'models', self.name)

    def is_installed(self):
        """
        Check if model (.pt file) is installed under SCT directory
        """
        if os.path.exists(os.path.join(self.folder, self.name + '.pt')) and \
                os.path.exists(os.path.join(self.folder, self.name + '.json')):
            return True
        else:
            raise FileNotFoundError("The model is not properly installed. Both the .pt and .json files should be "
                                    "present, and the basename should be the same as the folder name. Example: "
                                    "my_model/my_model.pt, my_model/my_model.json")

    def install(self):
        """
        Download and install model under SCT directory
        :param name_model:
        :return:
        """
        raise NotImplementedError

MODELS = [
        DeepsegModel(
                name = 'coord-t2star',
                url = 'https://osf.io/v9hs8/download?version=1',
                description = 'Cord segmentation on T2*-weighted contrast.'),
        DeepsegModel(
                name = 'uqueensland-mice-sc',
                url = 'https://osf.io/nu3ma/download?version=1',
                description = 'Cord segmentation on mouse MRI. Data from University of Queensland.'),
        DeepsegModel(
                name = 'uqueensland-mice-gm',
                url = 'https://osf.io/mfxwg/download?version=1'
                description = 'Gray matter segmentation on mouse MRI. Data from University of Queensland.'),
                ]
MODELS = {model.name: model for model in MODELS}
```

The other option is to operate directly on the dictionary, and in that case you don't even need the extra layer of a class. In this, users would simply say `import sct.models; models.is_installed('mouse-mri-2') ...`

```suggestion
MODELS = {
    'cord-t2star':
        {'url': 'https://osf.io/v9hs8/download?version=1',
         'description': 'Cord segmentation on T2*-weighted contrast.'},
    'uqueensland-mice-sc':
        {'url': 'https://osf.io/nu3ma/download?version=1',
         'description': 'Cord segmentation on mouse MRI. Data from University of Queensland.'},
    'uqueensland-mice-gm':
        {'url': 'https://osf.io/mfxwg/download?version=1',
         'description': 'Gray matter segmentation on mouse MRI. Data from University of Queensland.'},
    }

def folder(name):
    return os.path.join(__sct_dir__, 'models', name)
    
def is_installed(name):
    """
    Check if model (.pt file) is installed under SCT directory
    """
    if os.path.exists(os.path.join(folder(name), name + '.pt')) and \
            os.path.exists(os.path.join(folder(name), self.name + '.json')):
        return True
    else:
        raise FileNotFoundError("The model is not properly installed. Both the .pt and .json files should be "
                                "present, and the basename should be the same as the folder name. Example: "
                                "my_model/my_model.pt, my_model/my_model.json")

def install(name):
    """
    Download and install model under SCT directory
    :param name_model:
    :return:
    """
    raise NotImplementedError
```

Reply:
There's also this idiom I am fond of, not because it's a good idea but just because it's clever and has nice visual aesthetics: use the classes themselves as (singleton) data. It's a bit tricky and forces model names to follow python's naming rules:

```suggestion
class DeepsegModelMeta(type):
    @property
    def name(cls):
        return cls.__name__
    
    @property
    def folder(cls):
        return os.path.join(__sct_dir__, 'models', cls.name)

class DeepsegModel(metaclass=DeepsegModelMeta):
    """
    Template for metadata of packaged deep learning models.
    """
    name = None
    url = None
    description = None
    
    def is_installed(self):
        """
        Check if model (.pt file) is installed under SCT directory
        """
        if os.path.exists(os.path.join(self.folder, self.name + '.pt')) and \
                os.path.exists(os.path.join(self.folder, self.name + '.json')):
            return True
        else:
            raise FileNotFoundError("The model is not properly installed. Both the .pt and .json files should be "
                                    "present, and the basename should be the same as the folder name. Example: "
                                    "my_model/my_model.pt, my_model/my_model.json")
    
    def install(self):
        """
        Download and install model under SCT directory
        :param name_model:
        :return:
        """
        raise NotImplementedError
        
class coord_t2star(DeepsegModel):
        url = 'https://osf.io/v9hs8/download?version=1'
        description = 'Cord segmentation on T2*-weighted contrast.'

class uqueensland_mice_sc(DeepsegModel):
        url = 'https://osf.io/nu3ma/download?version=1'
        description = 'Cord segmentation on mouse MRI. Data from University of Queensland.'

class uqueensland_mice_gm(DeepsegModel):
        url = ' https://osf.io/mfxwg/download?version=1'
        description = 'Gray matter segmentation on mouse MRI. Data from University of Queensland.'

MODELS = {model.name: model for model in DeepsegModel.__subclasses__()}
```

Reply:
Great!